web/src/features/AppVersionHistory/api/getVersions.jsx

```diff
@@ -0,0 +1,197 @@
+      version.sequence === downstream?.currentVersion?.sequence;
+    const isDeploying = version.status === "deploying";
+    const isPastVersion = !!downstream?.pastVersions?.find(
+      (downstreamVersion) => downstreamVersion.sequence === version.sequence
+    );
+    const needsConfiguration = version.status === "pending_config";
+    const isRollback =
+      isPastVersion && version.deployedAt && currentApp?.allowRollback;
+    const isRedeploy =
+      isCurrentVersion &&
+      (version.status === "failed" || version.status === "deployed");
+    const canUpdateKots =
+      version.needsKotsUpgrade && !metadata?.isAirgap && !metadata?.isKurl;
+
+    let statusLabel = "";
+
+    if (needsConfiguration) {
+      statusLabel = "Configure";
+    } else if (downstream?.currentVersion?.sequence == undefined) {
+      if (canUpdateKots) {
+        statusLabel = "Upgrade";
+      } else {
+        statusLabel = "Deploy";
+      }
+    } else if (isRedeploy) {
+      statusLabel = "Redeploy";
+    } else if (isRollback) {
+      statusLabel = "Rollback";
+    } else if (isDeploying) {
+      statusLabel = "Deploying";
+    } else if (isCurrentVersion) {
+      statusLabel = "Deployed";
+    } else {
+      if (canUpdateKots) {
+        statusLabel = "Upgrade";
+      } else {
+        statusLabel = "Deploy";
+      }
+    }
+    return {
+      ...version,
+      statusLabel,
+    };
+  });
+
+  return {
+    ...versions,
+    versionHistory,
+  };
+}
+
+// TODO: refactor this function so that the airgapped / nonairgapped are separate
+function getVersionsSelectorForAirgapped({ versions, currentApp, metadata }) {
+  return getVersionsSelectorForKotsManaged({ versions, currentApp, metadata });
+}
+
+function getVersionsSelectorForHelmManaged({ versions }) {
+  const deployedSequence = versions?.versionHistory?.find(
+    (v) => v.status === "deployed"
+  )?.sequence;
+
+  const versionHistory = versions?.versionHistory.map((version) => {
+    let statusLabel = "Redeploy";
+
+    if (version.sequence > deployedSequence) {
+      statusLabel = "Deploy";
+    }
+
+    if (version.sequence < deployedSequence) {
+      statusLabel = "Rollback";
+    }
+    return {
+      ...version,
+      statusLabel,
+    };
+  });
+
+  return {
+    ...versions,
+    versionHistory,
+  };
+}
+
+function chooseVersionsSelector({
+  isAirgap,
+  isKurl,
+  isHelmManaged,
+  _getVersionsSelectorForKotsManaged = getVersionsSelectorForKotsManaged,
+  _getVersionsSelectorForAirgapped = getVersionsSelectorForAirgapped,
+  _getVersionsSelectorForHelmManaged = getVersionsSelectorForHelmManaged,
+}) {
+  // if airgapped
+  if (isAirgap && isKurl) {
+    return _getVersionsSelectorForAirgapped;
+  }
+
+  // if helm managed
+  if (isHelmManaged) {
+    return _getVersionsSelectorForHelmManaged;
+  }
+
+  // if kots managed
+  return _getVersionsSelectorForKotsManaged;
+}
+
+function useVersions({
+  currentPage,
+  refetchInterval,
+  pageSize,
+  _getVersions = getVersions,
```

Comment:
out of curiosity.. what does adding the underscore do?

Reply:
It's just a convention for renaming imports injected as arguments. So rather than trying to mock an import during unit tests, I can pass it in directly as an argument. 

in other languages `_variable` conveys "this is private, don't use it" to other developers.